kvserver: use AdminSplit in splitTestRange

tbg opened the pull request:
`splitTestRange` was previously reaching into the store to finagle a
split. It is used by around a dozen tests. Prototyping around #72374
has shown that these tests frequently need patching up whenever we
adjust (improve) the store's replica handling.

This is a time suck and besides, we also want to be able to test
the Store from within the `kvserver` (not `kvserver_test`) package.
So if we can make that happen, and can use AdminSplit, that would
be preferrable.

AdminSplit requires them to run a (somewhat) distributed multi-range
transaction. A first split would hit a single range, but after that the
split batch hits at least two ranges (meta2 and splitKey), and so we
need nontrivial DistSender-like functionality. Splits are also
nontrivial distributed transactions and so we need a TxnCoordSender.
Experience suggests that it's better to use the "real thing" and to
make sure it's configurable enough to fit the use case, rather than
whipping up half-baked replacements.

Luckily, it turned out that DistSender and TxnCoordSender are already
up to the task, and this commit adopts them in
`createTestStoreWithoutStart`, and changes `splitTestRange` to use
`AdminSplit`.

Release note: None

tbg committed:
kvserver: use AdminSplit in splitTestRange

`splitTestRange` was previously reaching into the store to finagle a
split. It is used by around a dozen tests. Prototyping around #72374
has shown that these tests frequently need patching up whenever we
adjust (improve) the store's replica handling.

This is a time suck and besides, we also want to be able to test
the Store from within the `kvserver` (not `kvserver_test`) package.
So if we can make that happen, and can use AdminSplit, that would
be preferrable.

AdminSplit requires them to run a (somewhat) distributed multi-range
transaction. A first split would hit a single range, but after that the
split batch hits at least two ranges (meta2 and splitKey), and so we
need nontrivial DistSender-like functionality. Splits are also
nontrivial distributed transactions and so we need a TxnCoordSender.
Experience suggests that it's better to use the "real thing" and to
make sure it's configurable enough to fit the use case, rather than
whipping up half-baked replacements.

Luckily, it turned out that DistSender and TxnCoordSender are already
up to the task, and this commit adopts them in
`createTestStoreWithoutStart`, and changes `splitTestRange` to use
`AdminSplit`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72383)
<!-- Reviewable:end -->

erikgrinaker commented on pkg/kv/kvserver/replica_command.go:
```diff
@@ -330,6 +330,7 @@ func (r *Replica) adminSplitWithDescriptor(
 			// correct range.
 			if !kvserverbase.ContainsKey(desc, args.Key) {
 				l, _ := r.GetLease()
+				_ = kvserverbase.ContainsKey(desc, args.Key)
```
Why this extra call?

tbg reviewed (commented):
Wow sorry for posting such a junk PR! It's actually not all junk but I submitted a bunch of stuff that came from my initial attempts at getting splits to work. Removed these extra bits.

bors r=erikgrinaker

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72383#-:-Mnf8wCu4sCN48nZnGCO:bxqiwql)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[pkg/kv/kvserver/replica_command.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72383#-Mna-E8mupvg8eSWTgbQ-r1-333:-Mnf9QTI2KEsnc4IK2PO:b-c01s5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c30b264a2a6f49b7f621e815f9f388c4ec879/pkg/kv/kvserver/replica_command.go#L333)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Why this extra call?
</blockquote></details>

Detritus. At one point I was getting mismatch errors here and I wanted to know why. Since `rr` doesn't work on my platform the easiest way to be able to step into `ContainsKey` with the right args is to duplicate it here. Removed.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
Yeah, was wondering about the DistSender emulation in the mock senders and stuff, but figured you'd added it temporarily for a reason. Nice to see it gone. :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72383#-:-MnfG81dCPQbHv0v6hY9:bfz1d4m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

tbg commented:
bors r-

Apparently it was needed for something? Urgh, going to take a look.

tbg commented:
Oh ok that makes sense. I fixed up the tests that use `createTestStore`, but there are tests using `testContext` that also call `splitTestRange`. So I need to do the thing I wanted to do anyway, which is have `testContext` use the real distributed txn stack as well. Okay, let's see if I can do that.

tbg committed:
kvserver: extract splits logic in createTestStoreWithoutStart

Release note: None

tbg committed:
kvserver: give unused keyspace to TestGCQueueProcess

The test use key prefixes that collide with kvs from cluster bootstrap
if the replica were properly bootstrapped. This will be the case in a
few commits, so give this test its own prefix to operate under.

Release note: None

tbg committed:
kvserver: remove two tests related to non-gossip of system config

These tests work today because the store they're operating under
is strung together and is missing lots of moving parts. Once we
use a more fully initialized stores these tests are prime candidates
for being flaky.

The real solution would be to to add testing knobs so that we could
investigate, in the test, the reasons for gossip, and to make sure
they are all valid reasons. But we are unlikely to break anything
in this area and are going to remove gossip of the system config
altogether in https://github.com/cockroachdb/cockroach/issues/70560.

So just remove these tests, which prevents them from getting in the
way of refactors of the test harness.

Release note: None

tbg committed:
kvserver: improve TestReplicaGossipConfigsOnLease

The old code would fail on a properly initialized store where there are
dozens of values in the initial system config.

Release note: None

tbg committed:
kvserver: improve TestReplicaLoadSystemConfigSpanIntent

Once the store is properly initialized, the system config will have
additional entries.

Release note: None

tbg committed:
storage,localtestcluster: disable randomized separated intents

Separated intents are fully baked in now, so this randomization makes
less sense. We continue to randomize in some tests, such as
TestMVCCHistories, so some coverage is retained.

The randomization caused nondeterminism during refactors of testContext
(where it was then affecting tests that were written with separated
intents hard-coded to "on") and so there was an immediate reason to
remove it.

Release note: None

tbg committed:
kvserver: use createTestStore in testContext [to be amended into previous commit]

This was a very satisfying cleanup. The commit history won't pass CI on
the individual commits, so this commit and all before it need to be
flattened. I'll do this once approved to make reviewing easier as the
changes are mostly non-overlapping (but they overlap through
`splitTestRange`).

Release note: None

tbg commented:
Pretty satisfying - got wrapped up with this just as I'm brushing up against the end of my day. I think this is fit for review though some small thing will certainly be wrong with it. `testContext.StartWithStoreConfigAndVersion` has become a very thin wrapper around `createTestStoreWithoutStart`, which I am quite happy about.

tbg committed:
kvserver: trim down TestStoreInitAndBootstrap

This test was duplicating parts of `createTestStoreWithoutStart` and it
was the one bogus hold-out user of `testSenderFactory`, which we now
get to remove.

Release note: None

erikgrinaker reviewed (approved):
LGTM at a glance. CI seems green too.

tbg commented:
tftr!

bors r=erikgrinaker

tbg commented:
I will stress kvserver before trying again. Pretty sure I've introduced a
few flakes.

(sent from mobile, please excuse my brevity)

On Sat, Nov 6, 2021, 15:37 craig[bot] ***@***.***> wrote:

> Build failed:
>
>    - GitHub CI (Cockroach)
>    <https://teamcity.cockroachdb.com/viewLog.html?buildId=3692359&buildTypeId=Cockroach_UnitTests>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/72383#issuecomment-962460926>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABGXPZEX6BOA4O2DOSW47CTUKU4UDANCNFSM5HI55PEQ>
> .
>

tbg committed:
kvserver: deflake TestReplicaGossipConfigsOnLease

It wasn't true that the first gossiped config we'd see would be the one
triggered by the lease.

Interestingly, I found that this test was flaky on master as well but it
isn't [detected] during nightly stress:

```
=== RUN   TestReplicaGossipConfigsOnLease
    test_log_scope.go:79: test logs captured to: /tmp/logTestReplicaGossipConfigsOnLease261191452
    test_log_scope.go:80: use -show-logs to present logs inline
    replica_test.go:1204: unexpected gossip of system config: values:<key:"\222" value:<raw_bytes:"\000\000\000\000\001T" timestamp:<> > >
    replica_test.go:1253: -- test log scope end --
test logs left over in: /tmp/logTestReplicaGossipConfigsOnLease261191452
--- FAIL: TestReplicaGossipConfigsOnLease (0.03s)
```

I checked the last [nightly] and the package did get stressed, with 20
iterations passing. Somewhat odd that I can reproduce this with a
targeted stress pretty much instantly, and it's not showing up for
years in CI.

[detected]: https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+TestReplicaGossipConfigsOnLease+is%3Aclosed
[nightly]: https://teamcity.cockroachdb.com/viewLog.html?buildId=3693551&buildTypeId=Cockroach_Nightlies_Stress&tab=buildLog&_focus=383#_state=226

Release note: None

tbg commented:
I went through 200 iters of the kvserver package (thx roachprod-stress) without issues. The one flake I saw during CI is now fixed; it was also flaky on master before.

bors r=erikgrinaker

tbg commented:
bors r-

ha, overlooked the "21 failures" line of my nonstandard invocation of `make roachprod-stress` 🙈

tbg commented:
bors r=erikgrinaker
80 runs so far, 0 failures, over 32m20s
I hadn't passed `-p` previously, so I was overloading the machines (probably).
